src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSecurityUtility.java

```diff
@@ -331,16 +341,29 @@ static void verifyColumnMasterKeyMetadata(SQLServerConnection connection, SQLSer
      */
     static SqlAuthenticationToken getManagedIdentityCredAuthToken(String resource,
             String managedIdentityClientId) throws SQLServerException {
-        ManagedIdentityCredential mic = null;
 
         if (logger.isLoggable(java.util.logging.Level.FINEST)) {
             logger.finest("Getting Managed Identity authentication token for: " + managedIdentityClientId);
         }
 
-        if (null != managedIdentityClientId && !managedIdentityClientId.isEmpty()) {
-            mic = new ManagedIdentityCredentialBuilder().clientId(managedIdentityClientId).build();
-        } else {
-            mic = new ManagedIdentityCredentialBuilder().build();
+        String key = getHashedSecret(new String[]{managedIdentityClientId, ManagedIdentityCredential.class.getSimpleName()});
+        ManagedIdentityCredential mic = (ManagedIdentityCredential) getCredentialFromCache(key);
+
+        if (null == mic) {
+            LOCK.lock();
+            ManagedIdentityCredentialBuilder micBuilder = new ManagedIdentityCredentialBuilder();
+
+            try {
+                if (null != managedIdentityClientId && !managedIdentityClientId.isEmpty()) {
+                    mic = micBuilder.clientId(managedIdentityClientId).build();
+                    addCredentialToCache(key, mic);
+                } else {
+                    mic = micBuilder.build();
+                    addCredentialToCache(key, mic);
+                }
+            } finally {
+                LOCK.unlock();
+            }

```

Comment:
Need to check the cache again after obtaining the lock since another thread might have populated the cache while we were waiting for the lock.
```suggestion
            try {
                ManagedIdentityCredentialBuilder micBuilder = (ManagedIdentityCredential) getCredentialFromCache(key);
                if (null == mic) {
                    micBuilder = new ManagedIdentityCredentialBuilder();
                    if (null != managedIdentityClientId && !managedIdentityClientId.isEmpty()) {
                        mic = micBuilder.clientId(managedIdentityClientId).build();
                        addCredentialToCache(key, mic);
                    } else {
                        mic = micBuilder.build();
                        addCredentialToCache(key, mic);
                    }
                }
            } finally {
                  LOCK.unlock();
            }
```